lib/web_git/diff.rb

```diff
@@ -0,0 +1,233 @@
+              file_content += "\n"
+            end
+          end
+          rights[filenames[i]] = file_content.chomp "\n"
+        end
+        rights
+      end
+
+      def self.get_last_commit_hash
+        Dir.chdir(Rails.root) do
+          log = `git log -1 --oneline`
+          log.split.first
+        end
+      end
+
+      def self.file_diffs(diff)
+        diff.scan(/diff --git.*?(?=diff --git|\z)/m)
+      end
+
+      def self.match_other_files(line, file, filenames)
+        filenames.each do |other_file|
+          if file != other_file
+            # It looks like:
+            #   --- a/<path-to-file>
+            #   +++ b/<path-to-file>
+            if line.include?('diff --git a/' + other_file + ' b/' + other_file)
+              return true
+            end
+          end
+        end
+        false
+      end
+
+      def self.get_file_names(commit)
+        Dir.chdir(Rails.root) do
+          if commit.blank?
+            filenames = `git diff --name-only`
+          else
+            filenames = `git diff-tree --no-commit-id --name-only -r #{commit}`
+          end
+          filenames.split("\n")
+        end
+      end
+
+      def self.get_last_diff
+        Dir.chdir(Rails.root) do
+          `git diff -M HEAD~1`
+        end
+      end
+
+      def self.get_last_left(diff)
+        filenames = get_file_names(get_last_commit_hash)
+        files = file_diffs(diff)
+        ones = {}
+        files.each_with_index do |file, i|
+          file_content = ""
+          lines = file.split("\n").drop(4)
+          lines.each do |line|
+            # The line numbers in the output of a git diff match this regex
+            # @@ -61,18 +61,15 @@
+            if !line.match?(/@@ ([-]\d+,\d+\s[+]\d+,\d+) @@/)
+              if line.first != "+"
+                line.slice!(0)
+                file_content += line + "\n"
+              end
+            else
+              file_content += "\n"
+            end
+          end
+          ones[filenames[i]] = file_content.chomp "\n"
+        end
+        ones
+      end
+
+      def self.get_last_right(diff)
+        filenames = get_file_names(get_last_commit_hash)
+        files = file_diffs(diff)
+        ones = {}
+        files.each_with_index do |file, i|
+          file_content = ""
+          lines = file.split("\n").drop(4)
+          lines.each do |line|
+            if !line.match?(/@@ ([-]\d+,\d+\s[+]\d+,\d+) @@/)
+              if line.first != "+"
+              elsif line.first == "+"
+                line.slice!(0)
+                file_content += line + "\n"
+              end
+            else
+              file_content += "\n"
+            end
+          end
+          ones[filenames[i]] = file_content.chomp "\n"
+        end
+        ones
+      end
+
+      def self.last_to_html(diff)
+        left_hash = get_last_left(diff)
+        right_hash = get_last_right(diff)
+
+        html_output = '<link rel="stylesheet"' +
```

Comment:
Since we're no longer using an iframe, perhaps all of the assets and things should just be in the pipeline or wherever the rest of the app's assets are coming from?